nifi-nar-bundles/nifi-snmp-bundle/nifi-snmp-processors/src/main/java/org/apache/nifi/snmp/processors/SetSNMP.java

```diff
@@ -34,191 +25,101 @@
-     */
     @Override
-    protected void onTriggerSnmp(ProcessContext context, ProcessSession processSession) throws ProcessException {
-        FlowFile flowFile = processSession.get();
+    public void onTrigger(final ProcessContext context, final ProcessSession processSession) {
+        final FlowFile flowFile = processSession.get();
         if (flowFile != null) {
-            // Create the PDU object
-            PDU pdu = null;
-            if(this.snmpTarget.getVersion() == SnmpConstants.version3) {
-                pdu = new ScopedPDU();
-            } else {
-                pdu = new PDU();
-            }
-            if(this.addVariables(pdu, flowFile.getAttributes())) {
-                pdu.setType(PDU.SET);
-                try {
-                    ResponseEvent response = this.targetResource.set(pdu);
-                    if(response.getResponse() == null) {
-                        processSession.transfer(processSession.penalize(flowFile), REL_FAILURE);
-                        this.getLogger().error("Set request timed out or parameters are incorrect.");
-                        context.yield();
-                    } else if(response.getResponse().getErrorStatus() == PDU.noError) {
-                        flowFile = SNMPUtils.updateFlowFileAttributesWithPduProperties(pdu, flowFile, processSession);
-                        processSession.transfer(flowFile, REL_SUCCESS);
-                        processSession.getProvenanceReporter().send(flowFile, this.snmpTarget.getAddress().toString());
-                    } else {
-                        final String error = response.getResponse().getErrorStatusText();
-                        flowFile = SNMPUtils.addAttribute(SNMPUtils.SNMP_PROP_PREFIX + "error", error, flowFile, processSession);
-                        processSession.transfer(processSession.penalize(flowFile), REL_FAILURE);
-                        this.getLogger().error("Failed while executing SNMP Set [{}] via " + this.targetResource + ". Error = {}", new Object[]{response.getRequest().getVariableBindings(), error});
-                    }
-                } catch (IOException e) {
-                    processSession.transfer(processSession.penalize(flowFile), REL_FAILURE);
-                    this.getLogger().error("Failed while executing SNMP Set via " + this.targetResource, e);
-                    context.yield();
-                }
-            } else {
+            try {
+                final SNMPResponse response = snmpRequestHandler.set(flowFile);
+                processResponse(processSession, flowFile, response);
+            } catch (SNMPException | IOException e) {
                 processSession.transfer(processSession.penalize(flowFile), REL_FAILURE);
-                this.getLogger().warn("No attributes found in the FlowFile to perform SNMP Set");
+                getLogger().error("Could not perform SNMP Set: Request timed out or parameters are incorrect.");
```

Comment:
Also based on some testing: the processor provides only a limited amount of information when the set is failed. It would be good to have more information on the interface. Also: I think timeout might be handled differently. It is possible "rolling back" the flow file and then yielding would be better, as possibly there is no issue with the input but we have a temporary outage in the other end